File: DSA/src/Questions/String/MergeAlternatively.java
package Questions.String;
/* 
 You are given two strings word1 and word2. Merge the strings by adding letters in alternating order, starting with word1. If a string is longer than the other, append the additional letters onto the end of the merged string.

Return the merged string.
Example 1:

Input: word1 = "abc", word2 = "pqr"
Output: "apbqcr"
Explanation: The merged string will be merged as so:
word1:  a   b   c
word2:    p   q   r
merged: a p b q c r
Example 2:

Input: word1 = "ab", word2 = "pqrs"
Output: "apbqrs"
Explanation: Notice that as word2 is longer, "rs" is appended to the end.
word1:  a   b 
word2:    p   q   r   s
merged: a p b q   r   s
Example 3:

Input: word1 = "abcd", word2 = "pq"
Output: "apbqcd"
Explanation: Notice that as word1 is longer, "cd" is appended to the end.
word1:  a   b   c   d
word2:    p   q 
merged: a p b q c   d
 

Constraints:

1 <= word1.length, word2.length <= 100
word1 and word2 consist of lowercase English letters.
 */
public class MergeAlternatively {
    public String mergeAlternately(String word1, String word2) {
        StringBuilder merged = new StringBuilder();
        int i = 0, j = 0;
        while (i < word1.length() && j < word2.length()) {
            merged.append(word1.charAt(i++));
            merged.append(word2.charAt(j++));
        }
        while (i < word1.length()) {
            merged.append(word1.charAt(i++));
        }
        while (j < word2.length()) {
            merged.append(word2.charAt(j++));
        }
        return merged.toString();
    }
    // Optimize the code by using StringBuilder instead of String
    // Time complexity: O(n)
    // Space complexity: O(n)

    public static void main(String[] args) {
        MergeAlternatively obj = new MergeAlternatively();
        String word1 = "abc";
        String word2 = "pqr";
        System.out.println(obj.mergeAlternately(word1, word2));
    }
}

File: DSA/Python/Questions/String/MergeAlternatively.py
"""
# Question: Merge Strings Alternately
# Link: https://leetcode.com/problems/merge-strings-alternately/

# Time Complexity: O(n + m)
# Space Complexity: O(n + m)

# Algorithm:
# 1. Alternate between strings
# 2. Track position in each string
# 3. Append remaining characters
# 4. Return merged string

# Key Components:
# - merge_alternately(): Main implementation
# - Two-pointer approach
# - String building
"""


class MergeAlternatively:
    def merge_alternately(self, word1: str, word2: str) -> str:
        merged = []
        i = j = 0

        # Merge alternately
        while i < len(word1) and j < len(word2):
            merged.append(word1[i])
            merged.append(word2[j])
            i += 1
            j += 1

        # Append remaining characters
        merged.extend(word1[i:])
        merged.extend(word2[j:])

        return "".join(merged)


def main():
    solution = MergeAlternatively()
    word1 = "abc"
    word2 = "pqr"
    print(solution.merge_alternately(word1, word2))  # Expected: "apbqcr"


if __name__ == "__main__":
    main()
